File: src/main/java/com/clinic/domain/entities/Appointment.java
package com.clinic.domain.entities;

import java.time.Instant;

/**
 * Core domain entity — mirrors the TypeScript "Appointment" entity from the AWS project. Pure
 * business object: no framework, no Azure, no persistence concerns. This is the heart of the Clean
 * Architecture domain layer.
 */
public class Appointment {

  private String appointmentId;
  private String insuredId;
  private int scheduleId;
  private CountryISO countryISO;
  private AppointmentStatus status;
  private Instant createdAt;
  private Instant completedAt;
  private Instant cancelledAt;
  private String contactEmail;

  public Appointment() {}

  public Appointment(
      String appointmentId, String insuredId, int scheduleId, CountryISO countryISO) {
    this.appointmentId = appointmentId;
    this.insuredId = insuredId;
    this.scheduleId = scheduleId;
    this.countryISO = countryISO;
    this.status = AppointmentStatus.PENDING;
    this.createdAt = Instant.now();
  }

  public void markCompleted() {
    if (this.status != AppointmentStatus.PENDING) {
      throw new IllegalStateException(
          "Only a PENDING appointment can be completed (current: " + this.status + ")");
    }
    this.status = AppointmentStatus.COMPLETED;
    this.completedAt = Instant.now();
  }

  public void markCancelled() {
    if (this.status != AppointmentStatus.PENDING) {
      throw new IllegalStateException(
          "Only a PENDING appointment can be cancelled (current: " + this.status + ")");
    }
    this.status = AppointmentStatus.CANCELLED;
    this.cancelledAt = Instant.now();
  }

  public void markRescheduled() {
    if (this.status != AppointmentStatus.PENDING) {
      throw new IllegalStateException(
          "Only a PENDING appointment can be rescheduled (current: " + this.status + ")");
    }
    this.status = AppointmentStatus.RESCHEDULED;
  }

  public String getAppointmentId() {
    return appointmentId;
  }

  public void setAppointmentId(String appointmentId) {
    this.appointmentId = appointmentId;
  }

  public String getInsuredId() {
    return insuredId;
  }

  public void setInsuredId(String insuredId) {
    this.insuredId = insuredId;
  }

  public int getScheduleId() {
    return scheduleId;
  }

  public void setScheduleId(int scheduleId) {
    this.scheduleId = scheduleId;
  }

  public CountryISO getCountryISO() {
    return countryISO;
  }

  public void setCountryISO(CountryISO countryISO) {
    this.countryISO = countryISO;
  }

  public AppointmentStatus getStatus() {
    return status;
  }

  public void setStatus(AppointmentStatus status) {
    this.status = status;
  }

  public Instant getCreatedAt() {
    return createdAt;
  }

  public void setCreatedAt(Instant createdAt) {
    this.createdAt = createdAt;
  }

  public Instant getCompletedAt() {
    return completedAt;
  }

  public void setCompletedAt(Instant completedAt) {
    this.completedAt = completedAt;
  }

  public Instant getCancelledAt() {
    return cancelledAt;
  }

  public void setCancelledAt(Instant cancelledAt) {
    this.cancelledAt = cancelledAt;
  }

  public String getContactEmail() {
    return contactEmail;
  }

  public void setContactEmail(String contactEmail) {
    this.contactEmail = contactEmail;
  }
}

File: clinic/domain/entities/appointment.py
"""Core domain entity - mirrors the TypeScript "Appointment" entity from the AWS project. Pure
business object: no framework, no Azure, no persistence concerns. This is the heart of the Clean
Architecture domain layer.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime

from clinic.domain.entities.appointment_status import AppointmentStatus
from clinic.domain.entities.country_iso import CountryISO
from clinic.domain.exceptions import IllegalStateError


@dataclass
class Appointment:
    appointment_id: str | None = None
    insured_id: str | None = None
    schedule_id: int = 0
    country_iso: CountryISO | None = None
    status: AppointmentStatus | None = None
    created_at: datetime | None = None
    completed_at: datetime | None = None
    cancelled_at: datetime | None = None
    contact_email: str | None = None

    @staticmethod
    def create(
        appointment_id: str, insured_id: str, schedule_id: int, country_iso: CountryISO
    ) -> Appointment:
        return Appointment(
            appointment_id=appointment_id,
            insured_id=insured_id,
            schedule_id=schedule_id,
            country_iso=country_iso,
            status=AppointmentStatus.PENDING,
            created_at=datetime.now(UTC),
        )

    def mark_completed(self) -> None:
        if self.status != AppointmentStatus.PENDING:
            raise IllegalStateError(
                f"Only a PENDING appointment can be completed (current: {self.status})"
            )
        self.status = AppointmentStatus.COMPLETED
        self.completed_at = datetime.now(UTC)

    def mark_cancelled(self) -> None:
        if self.status != AppointmentStatus.PENDING:
            raise IllegalStateError(
                f"Only a PENDING appointment can be cancelled (current: {self.status})"
            )
        self.status = AppointmentStatus.CANCELLED
        self.cancelled_at = datetime.now(UTC)

    def mark_rescheduled(self) -> None:
        if self.status != AppointmentStatus.PENDING:
            raise IllegalStateError(
                f"Only a PENDING appointment can be rescheduled (current: {self.status})"
            )
        self.status = AppointmentStatus.RESCHEDULED
